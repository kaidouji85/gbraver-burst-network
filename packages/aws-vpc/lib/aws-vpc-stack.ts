import { CfnOutput, Stack, StackProps } from "aws-cdk-lib";
import { aws_ec2 as ec2 } from "aws-cdk-lib";
import { Construct } from "constructs";

/** Cloud Formation Output プレフィックス */
const CFN_OUTPUT_PREFIX = 'VpcV2';

/** VPCスタックプロパティ */
interface VPCProps extends StackProps {
  /** サービス名 */
  service: string;
  /** VPC CIDR */
  cidr: string;
  /** AZ数 */
  maxAzs: number;
}

/** VPC スタック */
export class AwsVpcStack extends Stack {
  /**
   * @constructor
   * @param scope スコープ
   * @param id スタックのID
   * @param props CDKプロパティ
   */
  constructor(scope: Construct, id: string, props: VPCProps) {
    super(scope, id, props);
    const vpc = new ec2.Vpc(this, "vpc", {
      ipAddresses: ec2.IpAddresses.cidr(props.cidr),
      maxAzs: props.maxAzs,
      natGateways: 0,
    });
    new CfnOutput(this, "VpcId", {
      value: vpc.vpcId,
      exportName: `${props.service}:${CFN_OUTPUT_PREFIX}:VpcId`,
    });

    vpc.publicSubnets.forEach((subnets, index) => {
      new CfnOutput(this, `PublicNetAvailabilityZone${index}`, {
        value: subnets.availabilityZone,
        exportName: `${props.service}:${CFN_OUTPUT_PREFIX}:PublicNetAvailabilityZone${index}`,
      });
      new CfnOutput(this, `PublicSubnetId${index}`, {
        value: subnets.subnetId,
        exportName: `${props.service}:${CFN_OUTPUT_PREFIX}:PublicSubnetId${index}`,
      });
    });
  }
}
