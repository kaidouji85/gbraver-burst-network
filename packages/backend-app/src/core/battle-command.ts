import { Command, CommandSchema } from "gbraver-burst-core";
import { z } from "zod";

import { BattleID, BattleIDSchema, FlowID, FlowIDSchema } from "./battle";
import { UserID, UserIDSchema } from "./user";

/** バトルコマンド */
export type BattleCommand = {
  /** ユーザID */
  userID: UserID;
  /** バトルID */
  battleID: BattleID;
  /** フローID */
  flowID: FlowID;
  /** コマンド */
  command: Command;
};

/** BattleCommand zodスキーマ */
export const BattleCommandSchema = z.object({
  userID: UserIDSchema,
  battleID: BattleIDSchema,
  flowID: FlowIDSchema,
  command: CommandSchema,
});