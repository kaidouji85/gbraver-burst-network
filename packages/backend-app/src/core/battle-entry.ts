import type { ArmdozerId, PilotId } from "gbraver-burst-core";

import type { UserID } from "./user";
import type { WSAPIGatewayConnectionId } from "./ws-api-gateway-connection";

/**
 * ネット対戦エントリ
 * カジュアルマッチ、プライベートマッチ等、全ケースのエントリで
 * 必要最小限の情報をあつめた
 */
export type BattleEntry = {
  /** エントリーするユーザのID */
  userID: UserID;

  /** 選択したアームドーザID */
  armdozerId: ArmdozerId;

  /** 選択したパイロットID */
  pilotId: PilotId;

  /** コネクションID */
  connectionId: WSAPIGatewayConnectionId;
};
