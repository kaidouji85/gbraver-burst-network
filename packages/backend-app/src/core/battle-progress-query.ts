import { BattleID, FlowID } from "./battle";

/** バトル進行 問合わせパラメータ */
export type BattleProgressQuery = {
  /** バトルID */
  battleID: BattleID;
  /** フローID */
  flowID: FlowID;
};