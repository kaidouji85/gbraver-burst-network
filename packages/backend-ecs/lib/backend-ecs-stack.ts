import { Stack, StackProps } from "aws-cdk-lib";
import {
  aws_ec2 as ec2,
  aws_ecr as ecr,
  aws_ecs as ecs,
  aws_iam as iam,
  Duration,
} from "aws-cdk-lib";
import { Construct } from "constructs";

/** バックエンドECSスタックのプロパティ */
interface BackendEcsProps extends StackProps {
  /** サービス名 */
  service: string;
  /** ステージ名 */
  stage: string;
  /** 本ECSを起動するVPCのID */
  vpcId: string;
  /** サブネットのAZ */
  subnetAzs: string[];
  /** PublicサブネットのID */
  publicSubnetIds: string[];
  /** Websocket API GatewayのID */
  websocketAPIID: string;
  /** DynamoDB connections テーブルのARN */
  connectionsTableARN: string;
  /** DynamoDB casualMatchEntries テーブルのARN */
  casualMatchEntriesTableARN: string;
  /** DynamoDB battles テーブルのARN */
  battlesTableARN: string;
  /** マッチメイクECRリポジトリ名 */
  matchMakeEcrRepositoryName: string;
  /** Dockerイメージのタグ */
  dockerImageTag: string;
}

/** バックエンドECS スタック */
export class BackendEcsStack extends Stack {
  /**
   * @constructor
   * @param scope スコープ
   * @param id スタックID
   * @param props スタックのプロパティ
   */
  constructor(scope: Construct, id: string, props: BackendEcsProps) {
    super(scope, id, props);

    const vpc = ec2.Vpc.fromVpcAttributes(this, "backend-ecs-vpc", {
      vpcId: props.vpcId,
      availabilityZones: props.subnetAzs,
      publicSubnetIds: props.publicSubnetIds,
    });

    const matchMakePolicy = new iam.PolicyDocument({
      statements: [
        new iam.PolicyStatement({
          resources: [
            props.connectionsTableARN,
            props.casualMatchEntriesTableARN,
            props.battlesTableARN,
          ],
          actions: [
            "dynamodb:PutItem",
            "dynamodb:GetItem",
            "dynamodb:DeleteItem",
            "dynamodb:Scan",
            "dynamodb:BatchWrite*",
          ],
        }),
        new iam.PolicyStatement({
          resources: ["arn:aws:execute-api:*:*:**/@connections/*"],
          actions: ["execute-api:ManageConnections"],
        }),
      ],
    });
    const matchMakeServiceTaskRole = new iam.Role(
      this,
      "match-make-service-task-role",
      {
        assumedBy: new iam.ServicePrincipal("ecs-tasks.amazonaws.com"),
        inlinePolicies: { matchMakePolicy },
      }
    );
    const matchMakeTaskDefinition = new ecs.TaskDefinition(
      this,
      ",match-make-taskdef",
      {
        compatibility: ecs.Compatibility.FARGATE,
        cpu: "256",
        memoryMiB: "512",
        taskRole: matchMakeServiceTaskRole,
      }
    );
    const matchMakeLogging = new ecs.AwsLogDriver({
      streamPrefix: `${props.service}__${props.stage}__match-make`,
    });
    const matchMakeRepository = ecr.Repository.fromRepositoryName(
      this,
      "match-make-ecr",
      props.matchMakeEcrRepositoryName
    );
    matchMakeTaskDefinition.addContainer(`match-make-container`, {
      image: ecs.ContainerImage.fromEcrRepository(
        matchMakeRepository,
        props.dockerImageTag
      ),
      environment: {
        SERVICE: props.service,
        STAGE: props.stage,
        WEBSOCKET_API_ID: props.websocketAPIID,
      },
      logging: matchMakeLogging,
      healthCheck: {
        command: ["CMD-SHELL", "test -f match-make-health-check || exit 1"],
        interval: Duration.seconds(5),
        retries: 3,
      }
    });
    const cluster = new ecs.Cluster(this, "backend-ecs-cluster", { vpc });
    new ecs.FargateService(this, "service", {
      cluster,
      taskDefinition: matchMakeTaskDefinition,
      assignPublicIp: true,
    });
  }
}
