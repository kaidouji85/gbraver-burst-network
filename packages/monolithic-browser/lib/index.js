"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MonolithicBrowser = void 0;

var _login2 = require("./login");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

/** モノシリックサーバ ブラウザ用 SDK */
var MonolithicBrowser = /*#__PURE__*/function () {
  /**
   * コンストラクタ
   *
   * @param apiServerURL APIサーバURL
   */
  function MonolithicBrowser(apiServerURL) {
    _classCallCheck(this, MonolithicBrowser);

    this._apiServerURL = apiServerURL;
  }
  /**
   * ユーザID、パスワードでログインを行う
   * ログインに成功した場合はtrueを返す
   *
   * @param userID ユーザID
   * @param password パスワード
   * @return ログイン結果
   */


  _createClass(MonolithicBrowser, [{
    key: "login",
    value: function login(userID, password) {
      return (0, _login2.login)(userID, password, this._apiServerURL);
    }
  }]);

  return MonolithicBrowser;
}();

exports.MonolithicBrowser = MonolithicBrowser;