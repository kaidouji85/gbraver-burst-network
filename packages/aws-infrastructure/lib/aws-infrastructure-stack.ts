import * as cdk from '@aws-cdk/core';
import * as ec2 from "@aws-cdk/aws-ec2";
import * as ecs from "@aws-cdk/aws-ecs";
import * as ecr from "@aws-cdk/aws-ecr";
import * as iam from '@aws-cdk/aws-iam';

export class AwsInfrastructureStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const stage = new cdk.CfnParameter(this, 'stage', {
      type: 'String',
      description: 'serverless stage name',
    }).valueAsString;
    const websocketAPIID = cdk.Fn.importValue(`gbraver-burst-serverless:${stage}:WebsoketApiId`);
    const connectionsTableARN = cdk.Fn.importValue(`gbraver-burst-serverless:${stage}:ConnectionsTableArn`);
    const casualMatchEntriesTableARN = cdk.Fn.importValue(`gbraver-burst-serverless:${stage}:CasualMatchEntriesTableArn`);
    const battlesTableARN = cdk.Fn.importValue(`gbraver-burst-serverless:${stage}:BattlesTableArn`);

    const vpc = new ec2.Vpc(this, "vpc", {
      maxAzs: 1,
    });

    const matchMakeRepository = ecr.Repository.fromRepositoryName(this, "repo", "gbraver-burst-match-make");
    const matchMakePolicy = new iam.PolicyDocument({
      statements: [
        new iam.PolicyStatement({
          resources: [
            connectionsTableARN,
            casualMatchEntriesTableARN,
            battlesTableARN
          ],
          actions: [
            'dynamodb:PutItem',
            'dynamodb:GetItem',
            'dynamodb:DeleteItem',
            'dynamodb:Scan',
            'dynamodb:BatchWrite*',
          ],
        }),
        new iam.PolicyStatement({
          resources: ['arn:aws:execute-api:*:*:**/@connections/*'],
          actions: ['execute-api:ManageConnections'],
        }),
      ],
    });
    const matchMakeServiceTaskRole = new iam.Role(this, 'match-make-service-task-role', {
      roleName: 'ecs-service-task-role',
      assumedBy: new iam.ServicePrincipal('ecs-tasks.amazonaws.com'),
      inlinePolicies: {matchMakePolicy}
    });
    const matchMakeTaskDefinition = new ecs.TaskDefinition(this, "taskdef", {
      compatibility: ecs.Compatibility.FARGATE,
      cpu: "256",
      memoryMiB: "512",
      taskRole: matchMakeServiceTaskRole
    });
    const matchMakeLogging = new ecs.AwsLogDriver({
      streamPrefix: "gbraver-burst-match-make",
    });
    matchMakeTaskDefinition.addContainer(`match-make-container`, {
      image: ecs.ContainerImage.fromAsset('../serverless', {
        file: 'matchMake.Dockerfile'
      }),
      environment: {
        STAGE: stage,
        WEBSOCKET_API_ID: websocketAPIID,
      },
      logging: matchMakeLogging,
    });
    const cluster = new ecs.Cluster(this, "cluster", { vpc });
    const matchMakeService = new ecs.FargateService(this, "service", {
      cluster,
      taskDefinition: matchMakeTaskDefinition
    });
  }
}