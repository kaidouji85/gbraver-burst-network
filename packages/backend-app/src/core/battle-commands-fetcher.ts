import { BattlePlayer } from "./battle";
import { BattleCommand } from "./battle-command";

/** ゲーム参加プレイヤーのバトルコマンドを取得 */
export interface BattleCommandsFetcher {
  /**
   * ゲーム参加プレイヤーのバトルコマンドを取得する
   * @param players ゲーム参加プレイヤー
   * @return 取得結果、取得できなかった場合はnull
   */
  fetch(
    players: [BattlePlayer, BattlePlayer],
  ): Promise<[BattleCommand, BattleCommand] | null>;
}
