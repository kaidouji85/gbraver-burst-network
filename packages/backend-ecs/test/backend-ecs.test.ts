import { expect as expectCDK, matchTemplate, MatchStyle } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import {BackendEcsStack} from '../lib/backend-ecs-stack';

test('Empty Stack', () => {
    const app = new cdk.App();
    // WHEN
    const stack = new BackendEcsStack(app, 'MyTestStack', {
      stage: 'dev',
      vpcId: 'MyVpcId',
      privateSubnetId: 'MyPrivateSubnetId',
      privateNetAvailabilityZone: 'MyPrivateNetAvailabilityZone',
      websocketAPIID: 'MyWebsocketAPIID',
      connectionsTableARN: 'MyConnectionsTableARN',
      casualMatchEntriesTableARN: 'MyCasualMatchEntriesTableARN',
      battlesTableARN: 'MyBattlesTableARN',
    });
    // THEN
    expectCDK(stack).to(matchTemplate({
      "Resources": {}
    }, MatchStyle.EXACT))
});
