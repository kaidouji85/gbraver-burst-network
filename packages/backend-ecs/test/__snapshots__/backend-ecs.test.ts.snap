// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`バックエンドECSスタックがスナップショットと一致している 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "backendecscluster8F4B4819": Object {
      "Type": "AWS::ECS::Cluster",
    },
    "matchmakeservicetaskrole7B5E20D2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "dynamodb:PutItem",
                    "dynamodb:GetItem",
                    "dynamodb:DeleteItem",
                    "dynamodb:Scan",
                    "dynamodb:BatchWrite*",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    "MyConnectionsTableARN",
                    "MyCasualMatchEntriesTableARN",
                    "MyBattlesTableARN",
                  ],
                },
                Object {
                  "Action": "execute-api:ManageConnections",
                  "Effect": "Allow",
                  "Resource": "arn:aws:execute-api:*:*:**/@connections/*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "matchMakePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "matchmaketaskdef9DD39F0A": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Environment": Array [
              Object {
                "Name": "SERVICE",
                "Value": "gbraver-buesr-sls-dev",
              },
              Object {
                "Name": "STAGE",
                "Value": "v_1_5_1",
              },
              Object {
                "Name": "WEBSOCKET_API_ID",
                "Value": "MyWebsocketAPIID",
              },
            ],
            "Essential": true,
            "Image": Object {
              "Fn::Join": Array [
                "",
                Array [
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ".dkr.ecr.",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ".",
                  Object {
                    "Ref": "AWS::URLSuffix",
                  },
                  "/MyMatchMakeRepositoryName:v_1_5_1",
                ],
              ],
            },
            "LogConfiguration": Object {
              "LogDriver": "awslogs",
              "Options": Object {
                "awslogs-group": Object {
                  "Ref": "matchmaketaskdefmatchmakecontainertestuuidLogGroupB86F6A3E",
                },
                "awslogs-region": Object {
                  "Ref": "AWS::Region",
                },
                "awslogs-stream-prefix": "gbraver-buesr-sls-dev__v_1_5_1__match-make",
              },
            },
            "Name": "match-make-container-test-uuid",
          },
        ],
        "Cpu": "256",
        "ExecutionRoleArn": Object {
          "Fn::GetAtt": Array [
            "matchmaketaskdefExecutionRoleABBC5ED2",
            "Arn",
          ],
        },
        "Family": "MyTestStackmatchmaketaskdef1B2FC70A",
        "Memory": "512",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": Array [
          "FARGATE",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "matchmakeservicetaskrole7B5E20D2",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "matchmaketaskdefExecutionRoleABBC5ED2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "matchmaketaskdefExecutionRoleDefaultPolicy4AFB354E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ecr:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":repository/MyMatchMakeRepositoryName",
                  ],
                ],
              },
            },
            Object {
              "Action": "ecr:GetAuthorizationToken",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "matchmaketaskdefmatchmakecontainertestuuidLogGroupB86F6A3E",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "matchmaketaskdefExecutionRoleDefaultPolicy4AFB354E",
        "Roles": Array [
          Object {
            "Ref": "matchmaketaskdefExecutionRoleABBC5ED2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "matchmaketaskdefmatchmakecontainertestuuidLogGroupB86F6A3E": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "serviceSecurityGroupF051F0EB": Object {
      "Properties": Object {
        "GroupDescription": "MyTestStack/service/SecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": "MyVpcId",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "serviceService7DDC3B7C": Object {
      "Properties": Object {
        "Cluster": Object {
          "Ref": "backendecscluster8F4B4819",
        },
        "DeploymentConfiguration": Object {
          "MaximumPercent": 200,
          "MinimumHealthyPercent": 50,
        },
        "EnableECSManagedTags": false,
        "LaunchType": "FARGATE",
        "NetworkConfiguration": Object {
          "AwsvpcConfiguration": Object {
            "AssignPublicIp": "ENABLED",
            "SecurityGroups": Array [
              Object {
                "Fn::GetAtt": Array [
                  "serviceSecurityGroupF051F0EB",
                  "GroupId",
                ],
              },
            ],
            "Subnets": Array [
              "MyPrivateSubnetId",
            ],
          },
        },
        "TaskDefinition": Object {
          "Ref": "matchmaketaskdef9DD39F0A",
        },
      },
      "Type": "AWS::ECS::Service",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
